Sync historical beefy justifications on relayer start up

alistair-singh opened the pull request:
https://trello.com/c/g3iv7eFj/124-beefy-relayer-historical-sync

## Info
This PR implements the following workflow:
* At startup use the query api to check how far behind the relay is behind the latest block.
* Catch up those blocks.
* Resume subscription to beefy justifications.

## Testing
E2E tests pass so there is no regression in functionality. However the new sync code path is till being debugged and tested

## Notes
1. For this revision we use `BeefyStartingBlock` from the config as the last known block. The relayer needs to instead find the last block it knows about.
2. Originally `BlockCatchUpLimit` was meant to implement a mechanism to catch up 50 blocks at a time. However after looking at the data, we only know if a block has a justification by fetching it and we cannot skip blocks in between as we would miss justifications. In this PR it is set to `1` to check every block. If my assumption that we have to inspect every block is correct than this can be removed and replaced with running only running the sync if we are more than 50 blocks behind.
3. While debugging I noticed that the `Justification` field of `SignedBlock` is always nil. Hence the processing code is not yet fully tested. 
4. Possible edge case that is not catered for in this PR yet is that if the relayer is far behind and has to sync a lot of blocks, it could lead to a situation where by the time it has sync'd the backlog of blocks, it is still more than 50 blocks behind because new blocks have been created while processing.

musnit commented:
Looks like a good start!

> * For this revision we use `BeefyStartingBlock` from the config as the last known block. The relayer needs to instead find the last block it knows about.
yeh - we can do this in a later PR

> * Originally `BlockCatchUpLimit` was meant to implement a mechanism to catch up 50 blocks at a time. However after looking at the data, we only know if a block has a justification by fetching it and we cannot skip blocks in between as we would miss justifications. In this PR it is set to `1` to check every block. If my assumption that we have to inspect every block is correct than this can be removed and replaced with running only running the sync if we are more than 50 blocks behind.
You're right, we will need to inspect one block at a time to find a justification. However, when we find one, we can jump ahead. So I think the process can be something like
 - get starting block
 - jump ahead BlockCatchUpLimit
 - look one-block-at-a-time, find a justification, process it, jump ahead BlockCatchUpLimit, repeat

> * While debugging I noticed that the `Justification` field of `SignedBlock` is always nil. Hence the processing code is not yet fully tested.
Yeh - I did try my side and found some - maybe only 1 in5 or 1 in 10 has a justification. It may be worth playing with polkadotjs's UI to query blocks manually try find a specific block that has a justification so you can test that specific blocknumber to ensure there are no other issues resulting in the nil. If you still don't find one with further testing let me know and can look into if there is a potential issue.

> 4. Possible edge case that is not catered for in this PR yet is that if the relayer is far behind and has to sync a lot of blocks, it could lead to a situation where by the time it has sync'd the backlog of blocks, it is still more than 50 blocks behind because new blocks have been created while processing.
yeh, maybe instead of looping from startBlockNumber to endBlockNumber you could just loop until you query for a block that does not yet exist and use that as a stopping condition?

musnit commented:
Another concern: It is important that the justifications are processed in order when they get to Ethereum, otherwise they won't work. With us looping quickly through the blocks in the listener, we need to confirm:
 - blocks will be received and processed by the ethereum writer in the same order
 - We may also need to ensure that the writer submits them to Ethereum in order
 - We likely do also need to ensure that Ethereum accepts those transactions in order. Transactions sent to Ethereum do have an ordered nonce, so if the writer is incrementing its nonce correctly as it submits multiple transactions in a row then we should be fine, but should double check this.

alistair-singh commented:
Rebased latest changes from main into this branch.

> You're right, we will need to inspect one block at a time to find a justification. However, when we find one, we can jump ahead. So I think the process can be something like

> * get starting block
> * jump ahead BlockCatchUpLimit
> * look one-block-at-a-time, find a justification, process it, jump ahead BlockCatchUpLimit, repeat

The above algorithm is implemented.

> yeh, maybe instead of looping from startBlockNumber to endBlockNumber you could just loop until you query for a block that does not yet exist and use that as a stopping condition?

This has also been implemented. When you request for a block hash that does not exist an error 'required result to be 32 bytes, but got 0' gets returned. See [code](https://github.com/Snowfork/go-substrate-rpc-client/blob/main/types/hash.go#L87). This means that the only way to check for this case at this moment is string comparison which does not feel clean.

> Yeh - I did try my side and found some - maybe only 1 in5 or 1 in 10 has a justification. It may be worth playing with polkadotjs's UI to query blocks manually try find a specific block that has a justification so you can test that specific blocknumber to ensure there are no other issues resulting in the nil. If you still don't find one with further testing let me know and can look into if there is a potential issue.

The reason for this is that the tag on `Justification` is `json:"justification"` but in the response body from rpc call, the response json uses `justifications` instead. The format of the justification is also different from the json returned from the subscribe api. It is returned as `[][][]Byte`.

```sh
 curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "chain_getBlock", "params":["0x67752be6b16a08be2c601e992a470e8eecb7ac895c33867ae7f3d87b68395749"]}' http://localhost:9933
```
```json
"justifications":[
  [
    [66,69,69,70], // BEEF once converted to string
[1,53,143,75,74,43,11,73,32,26,52,189,152,142,114,92,57,80,154,75,149,13,106,195,208,106,225,119,99,172,90,54,190,177,3,0,0,0,0,0,0,0,0,0,0,12,1,93,188,186,6,10,93,49,146,187,97,3,184,97,26,24,18,193,183,74,43,207,42,21,28,152,127,175,66,203,188,84,80,65,249,76,100,69,168,200,8,103,188,204,101,71,228,115,161,106,53,145,55,224,88,228,15,94,42,61,140,255,86,38,6,0,1,241,45,60,223,168,67,201,96,107,136,143,87,151,244,103,87,188,112,109,49,129,112,1,184,65,112,130,192,163,208,95,173,101,142,246,117,159,198,95,173,142,187,119,23,79,63,89,114,13,145,251,204,2,216,135,49,91,181,88,243,230,101,254,77,1,1,253,12,20,142,103,129,86,124,127,244,80,50,40,190,65,158,64,151,208,64,114,35,91,207,232,163,250,207,136,107,123,217,39,104,8,89,201,34,46,241,161,128,38,201,204,165,46,228,104,48,202,215,213,6,6,140,218,164,142,63,247,5,97,47,1]
  ]
]
```

Currently the code can partially decode this format successfully. It is just missing signatures.

alistair-singh commented:
This PR depends on [another PR](https://github.com/Snowfork/go-substrate-rpc-client/pull/18) in `go-substrate-rpc-client`.

All feedback has been address except for:
> Another concern: It is important that the justifications are processed in order when they get to Ethereum, otherwise they won't work. With us looping quickly through the blocks in the listener, we need to confirm:

e2e tests pass.

alistair-singh commented on relayer/relays/beefy/beefy-relaychain-listener.go:
```diff
@@ -39,12 +40,82 @@ func NewBeefyRelaychainListener(
 func (li *BeefyRelaychainListener) Start(ctx context.Context, eg *errgroup.Group) error {
 
 	eg.Go(func() error {
+		// TODO: Get the starting block from some place other than config.
```
Added new issue. 
https://github.com/Snowfork/snowbridge/issues/506

alistair-singh commented on relayer/relays/beefy/beefy-relaychain-listener.go:
```diff
@@ -18,6 +17,8 @@ import (
 	log "github.com/sirupsen/logrus"
 )
 
+const SyncBlockNumberJump = 50
```
Added this to infra as well.
https://github.com/Snowfork/snowbridge-infra/pull/3

alistair-singh commented:
Addressed comments:
* structured logging
* potential deadlocking
* moved SyncBlockNumberJump to config

Rebased from `main` and e2e tests pass with latest changes.

vgeddes commented on relayer/relays/beefy/beefy-relaychain-listener.go:
```diff
@@ -39,12 +38,82 @@ func NewBeefyRelaychainListener(
+func (li *BeefyRelaychainListener) syncBeefyJustifications(ctx context.Context, startBlockNumber, syncBlockNumberJump uint64) error {
+	log.WithFields(log.Fields{"StartingBlockNumber": startBlockNumber, "SyncBlockNumberJump": syncBlockNumberJump}).Info("Syncing BEEFY justifications.")
+
+	blockNumber := startBlockNumber + syncBlockNumberJump
```
@musnit @alistair-singh I'm not familiar with why we are allowed to skip justifications, so I think adding a comment in the code explaining that would be very useful.

alistair-singh commented on relayer/relays/beefy/config.go:
```diff
@@ -10,7 +10,8 @@ type Config struct {
 type SourceConfig struct {
-	Polkadot config.PolkadotConfig `mapstructure:"polkadot"`
+	Polkadot            config.PolkadotConfig `mapstructure:"polkadot"`
+	SyncBlockNumberJump uint64                `mapstructure:"sync-block-number-jump"`
```
Yeah I didn't like this name very much either. `justification-skip-period` is better but `period` feels like its related to some time interval. How about `justification-skip-count`?

vgeddes replied:
`justification-skip-count` sounds fine :+1:

alistair-singh commented:
e2e tests are currently timing out for Substrate to Ethereum test cases. I think this is related to the fact that we are always skipping. I will change the code to skip only on initialization(i.e. have separate sync and sub flows again).
```
  1) Bridge
       ERC20 App
         should transfer ERC20 from Substrate to Ethereum:
     Error: Timeout of 360000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/alistair/snowbridge/test/test/erc20app.js)
      at listOnTimeout (internal/timers.js:554:17)
      at processTimers (internal/timers.js:497:7)

  2) Bridge
       ERC721 App
         should transfer ERC721 token from Substrate to Ethereum:
     Error: Timeout of 360000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/alistair/snowbridge/test/test/erc721app.js)
      at listOnTimeout (internal/timers.js:554:17)
      at processTimers (internal/timers.js:497:7)

  3) Bridge
       ETH App
         should transfer ETH from Substrate to Ethereum (incentivized channel):
     Error: Timeout of 360000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/alistair/snowbridge/test/test/ethapp.js)
      at listOnTimeout (internal/timers.js:554:17)
      at processTimers (internal/timers.js:497:7)

```

musnit commented:
> e2e tests are currently timing out for Substrate to Ethereum test cases. I think this is related to the fact that we are always skipping. I will change the code to skip only on initialization(i.e. have separate sync and sub flows again).
> 
> ```
>   1) Bridge
>        ERC20 App
>          should transfer ERC20 from Substrate to Ethereum:
>      Error: Timeout of 360000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/alistair/snowbridge/test/test/erc20app.js)
>       at listOnTimeout (internal/timers.js:554:17)
>       at processTimers (internal/timers.js:497:7)
> 
>   2) Bridge
>        ERC721 App
>          should transfer ERC721 token from Substrate to Ethereum:
>      Error: Timeout of 360000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/alistair/snowbridge/test/test/erc721app.js)
>       at listOnTimeout (internal/timers.js:554:17)
>       at processTimers (internal/timers.js:497:7)
> 
>   3) Bridge
>        ETH App
>          should transfer ETH from Substrate to Ethereum (incentivized channel):
>      Error: Timeout of 360000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/alistair/snowbridge/test/test/ethapp.js)
>       at listOnTimeout (internal/timers.js:554:17)
>       at processTimers (internal/timers.js:497:7)
> ```

I think a better fix would be to keep the poll/sync flow, but only skip if we are far behind. If we are close, don't skip.

alistair-singh commented:
e2e tests now pass however it feels like the ` Substrate to Ethereum` cases timeout more. Increasing the timeout makes the tests pass consistently. A single test takes between 190 and 400 seconds to complete. The current timeout is 280 seconds.

I have also run the e2e tests while killing the relayers every 2 minutes, to test graceful recovery, and they pass.

musnit commented:
awesome:
 - longer time out i guess is fine but 400 seems high. do you know why it's
a lot slower? the beefy-relayer has <10 seconds polling time so i would
think it shouldn't delay things by more than an extra 10 seconds (unless
you're killing it, ofc)

On Tue, Aug 31, 2021 at 6:46 PM Alistair Singh ***@***.***>
wrote:

> e2e tests now pass however it feels like the Substrate to Ethereum cases
> timeout more. Increasing the timeout makes the tests pass consistently. A
> single test takes between 190 and 400 seconds to complete. The current
> timeout is 280 seconds.
>
> I have also run the e2e tests while killing the relayers every 2 minutes,
> to test graceful recovery, and they pass.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Snowfork/snowbridge/pull/497#issuecomment-909765876>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFVTOKBI26LQATUYOYPSQLT7VZWRANCNFSM5CJH3WNA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

alistair-singh commented:
Closing this out as we as switched back to a hybrid approach.
